api4/user.go

```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
 	c.LogAudit("TermsOfServiceId=" + termsOfServiceId + ", accepted=" + strconv.FormatBool(accepted))
 	ReturnStatusOK(w)
 }
+
+func getUserAnalytics(c *Context, w http.ResponseWriter, r *http.Request) {
+	// Types of data sets to return
+	// All data sets are defined as Datasets in mattermost-redux/src/constants/user_analytics
+	// Data sets are mirrored in the struct, UserAnalytics, in model/user.go
+	types := r.URL.Query().Get("types")
```

Comment:
I originally pass it as an array, but it gets converted into a comma delimited list of strings during the query string building. I agree using .contains on a string kind of sucks. I now convert it back to an array of strings, and use the `utils.StringInSlice` function to check if the array contains a certain data set. I think that is better. What do you think?

Reply:
Since StringInSlice is linear time, I think the scaling advantage over contains is probably small. Better would be to build a set out of the list, which will have constant- or log-time contains. This is only true if the list will be "large" however, either now or in future. It's not clear to me what types actually is, so I'm not sure it will be large enough to merit this optimization.

Reply:
Ah makes sense. Is a set like an object? We are now just passing a 'name' variable for identifying one data set to return, so this has become redundant.

Reply:
A set is a data structure. It's a bit like a list, except that:

- Items must be unique (usually). If you try to put a duplicate in, it usually isn't added.
- To determine whether something is in a set is (usually) O(1) or O(log n) time (where n is the number of elements), whereas for a list it is O(n) time.

In this application, where you do O(n) contains queries, you get O(n) total runtime with a set, but O(n^2) with a list. For a list of a 1000, that would be a million extra ops, but for a list of 100, it'd be only 9000 extra ops. That's why I wasn't sure whether this optimization was worth thinking about.